internal/users/manager.go

```diff
@@ -292,33 +383,254 @@ func (m *Manager) AllShadows() ([]ShadowEntry, error) {
+
+	return uid, nil
 }
 
-// GenerateGID deterministically generates an ID between from the given string, ignoring case,
-// in the range [GIDMin, GIDMax]. The generated ID is *not* guaranteed to be unique.
-func (m *Manager) GenerateGID(str string) uint32 {
-	return generateID(str, m.config.GIDMin, m.config.GIDMax)
+// registerUser registers a temporary user with the given name and a unique UID in our NSS handler (in memory, not in
+// the database).
+//
+// The caller must lock m.temporaryEntriesMu for writing before calling this function.
+//
+// Returns the generated UID and a cleanup function that should be called to remove the temporary user once the user was
+// added to the database.
+func (m *Manager) registerUser(name string, preAuth bool) (uid uint32, cleanup func(), err error) {
+	// Check if there is already a temporary user for that name
+	for uid, user := range m.temporaryUsers {
+		if user.name == name+"-preauth" {
+			// A temporary user with the same name is already registered. To avoid that we generate multiple UIDs for
+			// the same user, we return the already generated UID.
+			cleanup = func() { m.deleteTemporaryUser(uid) }
+			return uid, cleanup, nil
+		}
+	}
+
+	for {
+		uid, err = m.GenerateUID()
+		if err != nil {
+			return 0, nil, err
+		}
+
+		// To avoid races where a user with this UID is created by some NSS source after we checked, we register this
+		// UID in our NSS handler and then check if another user with the same UID exists in the system. This way we
+		// can guarantee that the UID is unique, under the assumption that other NSS sources don't add users with a UID
+		// that we already registered (if they do, there's nothing we can do about it).
+		var tmpName string
+		tmpName, cleanup, err = m.addTemporaryUser(uid, preAuth)
+		if errors.Is(err, errUserAlreadyExists) {
+			log.Debugf(context.Background(), "UID %d already in use, generating a new one", uid)
+			continue
+		}
+		if err != nil {
+			return 0, nil, fmt.Errorf("could not register temporary user: %w", err)
+		}
+
+		if unique := m.checkIsUniqueUID(uid, tmpName); unique {
+			if preAuth {
+				// Rename the temporary user so that we can recognize it later.
+				tmpName = name + "-preauth"
+				m.temporaryUsers[uid] = temporaryUser{name: tmpName, preAuth: preAuth}
+			}
+			log.Debugf(context.Background(), "Registered temporary user %q with UID %d", tmpName, uid)
+			break
+		}
+
+		// If the UID is not unique, remove the temporary user and generate a new one in the next iteration.
+		cleanup()
+	}
+
+	return uid, cleanup, nil
 }
 
-func generateID(str string, minID, maxID uint32) uint32 {
-	str = strings.ToLower(str)
+// checkIsUniqueUID checks if the given UID is unique in the system. It returns false if the UID is already assigned to
+// a user by any NSS source (except the given temporary user) and true otherwise.
+func (m *Manager) checkIsUniqueUID(uid uint32, tmpName string) bool {
+	for _, entry := range localgroups.GetPasswdEntries() {
```

Comment:
`slices.ContainsFunc`?

Reply:
That wouldn't make this code any simpler, and IMO the loop is a bit more readable and concise:

```diff
diff --git a/internal/users/manager.go b/internal/users/manager.go
index 883b3d5e..6bd9c606 100644
--- a/internal/users/manager.go
+++ b/internal/users/manager.go
@@ -8,6 +8,7 @@ import (
        "fmt"
        "math/big"
        "os"
+       "slices"
        "sync"
        "syscall"
 
@@ -466,14 +467,13 @@ func (m *Manager) registerUser(name string, preAuth bool) (uid uint32, cleanup f
 // isUniqueUID returns true if the given UID is unique in the system. It returns false if the UID is already assigned to
 // a user by any NSS source (except the given temporary user).
 func (m *Manager) isUniqueUID(uid uint32, tmpName string) bool {
-       for _, entry := range localentries.GetPasswdEntries() {
+       return slices.ContainsFunc(localentries.GetPasswdEntries(), func(entry localentries.Passwd) bool {
                if entry.UID == uid && entry.Name != tmpName {
                        log.Debugf(context.Background(), "UID %d already in use by user %q, generating a new one", uid, entry.Name)
-                       return false
+                       return true
                }
-       }
-
-       return true
+               return false
+       })
 }
```